### PR TITLE
Run ensemble experiment with design matrix

### PR DESCRIPTION
**Pull request by xjules**
**Issue**
Resolves https://github.com/equinor/ert/issues/8961

**Approach**
- Do design matrix validation on startup
  - Validate active realizations and design matrix xsls
- Use design_matrix parameters in ensemble experiment
- add merge function to merge design parameters with existing parameters
  - currently if there is a full overlap the design matrix parameters takes precedence
  - we raise Validation error in case of partial overlap

 - [x] remove poly_design once done.
 - [x] add unit and cli tests

(Screenshot of new behavior in GUI if applicable)

When hitting run and overlapping fails:
image

When loading wrongly formatted design matrix:
image

- [x] PR title captures the intent of the changes, and is fitting for release notes.
- [x] Added appropriate release note label
- [ ] Commit history is consistent and clean, in line with the [contribution guidelines](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md).
- [x] Make sure unit tests pass locally after every commit (`git rebase -i main
      --exec 'pytest tests/ert/unit_tests -n logical -m "not integration_test"'`)

## When applicable
- [ ] **When there are user facing changes**: Updated documentation
- [ ] **New behavior or changes to existing untested code**: Ensured that unit tests are added (See [Ground Rules](https://github.com/equinor/ert/blob/main/CONTRIBUTING.md#ground-rules)).
- [ ] **Large PR**: Prepare changes in small commits for more convenient review
- [ ] **Bug fix**: Add regression test for the bug
- [ ] **Bug fix**: Create Backport PR to latest release

### CODE REVIEW

File: src/ert/gui/simulation/ensemble_experiment_panel.py [hunk 1]
```diff
@@ -85,18 +85,27 @@ def __init__(
...
             button_layout.addWidget(show_dm_param_button)
             button_layout.addStretch()  # Add stretch to push the button to the left
 
             layout.addRow("Design Matrix", button_layout)
             show_dm_param_button.clicked.connect(
                 lambda: self.on_show_dm_params_clicked(design_matrix)
             )
+            use_dm_checkbox.stateChanged.connect(
+                lambda: self.on_use_dm_checkbox_toggled(
+                    use_dm_checkbox.checkState(), design_matrix
```
Comment (jonathan-eq) [inline src/ert/gui/simulation/ensemble_experiment_panel.py:106]: I wonder if we have to do `use_dm_checkbox.checkState()` or if the new state is already given through the emitted signal, so we can include it as an argument to the lambda. Can we double check this?

Comment (xjules) [reply src/ert/gui/simulation/ensemble_experiment_panel.py:106]: I'm going to remove the checkbox as it is not really necessary I think.

File: src/ert/enkf_main.py [hunk 2]
```diff
@@ -148,6 +150,26 @@ def _seed_sequence(seed: Optional[int]) -> int:
     return int_seed
 
 
+def load_prior(
+    ensemble: Ensemble, active_realizations: Iterable[int], design_matrix: DesignMatrix
+) -> None:
+    assert design_matrix.parameter_configuration is not None
+    assert (
+        design_matrix.design_matrix_df is not None
```
Comment (dafeda) [inline src/ert/enkf_main.py:158]: Perhaps better to do in a different PR, but could we instead just call it `df` so that we get `design_matrix.df`?

File: tests/ert/ui_tests/cli/analysis/test_design_matrix.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+import pandas as pd
+import pytest
+
+from ert.config import ErtConfig
+from ert.mode_definitions import ENSEMBLE_EXPERIMENT_MODE
+from ert.storage import open_storage
+from tests.ert.ui_tests.cli.run_cli import run_cli
+
+
+@pytest.mark.usefixtures("copy_poly_case")
```
Comment (dafeda) [inline tests/ert/ui_tests/cli/analysis/test_design_matrix.py:13]: When I try running this test locally it fails with `Worksheet named 'DefaultSheet' not found`.
Should this use fixture "copy_poly_design_case" or something instead?

File: src/ert/enkf_main.py (hunk 2 above)
Comment (xjules) [reply src/ert/enkf_main.py:158]: we usually try to avoid using abbreviations but this one might be ok :+1:

File: tests/ert/ui_tests/cli/analysis/test_design_matrix.py (hunk 3 above)
Comment (xjules) [reply tests/ert/ui_tests/cli/analysis/test_design_matrix.py:15]: I have removed `test-data/poly_design` as we can generate this case on the fly and then it we need the matrix for experiments it can be copied from the temp dir.

Comment (xjules) [reply tests/ert/ui_tests/cli/analysis/test_design_matrix.py:15]: > When I try running this test locally it fails with `Worksheet named 'DefaultSheet' not found`.

This should be fixed now. Got a commit lost when rebasing but now it should be back.

File: src/ert/run_models/ensemble_experiment.py [hunk 4]
```diff
@@ -61,7 +61,34 @@ def run_experiment(
         restart: bool = False,
     ) -> None:
         self.log_at_startup()
-        if not restart:
+        # sensitivity analysis
```
Comment (xjules) [inline src/ert/run_models/ensemble_experiment.py:64]: Note to self: This comment is wrong

File: src/ert/enkf_main.py (hunk 2 above)
Comment (dafeda) [reply src/ert/enkf_main.py:158]: It's a common naming convention in pandas.

Comment (xjules) [reply src/ert/enkf_main.py:158]: I'm aware of df in pandas but not sure if all are. Anyway this can be changed in another pr.

File: src/ert/enkf_main.py [hunk 5]
```diff
@@ -148,6 +150,26 @@ def _seed_sequence(seed: Optional[int]) -> int:
     return int_seed
 
 
+def load_prior(
```
Comment (dafeda) [inline src/ert/enkf_main.py:153]: Could we come up with a more descriptive name for this function? `save_design_matrix_to_ensemble` perhaps?

Comment (xjules) [reply src/ert/enkf_main.py:153]: I wanted to be consistent with the sample_priors (which also save the params to ensemble afterwards), but I think it is better to be descriptive.

Comment (xjules) [reply src/ert/enkf_main.py]: I've changed it as suggested. I've also replaced DesignMatrix into pd.DataFrame directly. I think I need to write a test for it.

File: src/ert/run_models/ensemble_experiment.py [hunk 6]
```diff
@@ -61,7 +61,35 @@ def run_experiment(
...
-        if not restart:
+        # If design matrix is present, we substitute the experiment parameters
+        # with those in the design matrix
+        if self.ert_config.analysis_config.design_matrix is not None:
+            if (
+                self.ert_config.analysis_config.design_matrix.parameter_configuration
+                is None
+            ):
+                self.ert_config.analysis_config.design_matrix.read_design_matrix()
+            assert (
```
Comment (dafeda) [inline src/ert/run_models/ensemble_experiment.py:72]: Could we get rid of this assert and make `read_design_matrix` responsible for ensuring that `design_matrix.parameter_configuration` is not None?

Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:72]: mypy was complaining ...

File: src/ert/run_models/ensemble_experiment.py [hunk 7]
```diff
@@ -61,7 +61,35 @@ def run_experiment(
         restart: bool = False,
     ) -> None:
         self.log_at_startup()
-        if not restart:
+        # If design matrix is present, we substitute the experiment parameters
+        # with those in the design matrix
+        if self.ert_config.analysis_config.design_matrix is not None:
+            if (
+                self.ert_config.analysis_config.design_matrix.parameter_configuration
+                is None
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:69]: For example when you run ert from cli.

[the remark this answers is not in the document]
Comment (dafeda) [reply src/ert/run_models/ensemble_experiment.py:69]: I don't understand. Why is it different when running from cli?

[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:69]: Since, when you open gui, you can open experiment panel, which does the initialization. In cli, you just create run_model and run.

[the remark this answers is not in the document]
Comment (dafeda) [reply src/ert/run_models/ensemble_experiment.py:69]: Alright, thanks. Would be nice to add that as a comment since it's not obvious.

File: tests/ert/unit_tests/test_libres_facade.py [hunk 8]
```diff
@@ -241,3 +244,53 @@ def test_load_gen_kw_not_sorted(storage, tmpdir, snapshot):
...
+):
+    design_path = tmp_path / "design_matrix.xlsx"
+    ensemble_size = 10
+    a_values = np.random.default_rng().uniform(-5, 5, 10)
+    b_values = np.random.default_rng().uniform(-5, 5, 10)
+    c_values = np.random.default_rng().uniform(-5, 5, 10)
+    design_matrix_df = DataFrame({"a": a_values, "b": b_values, "c": c_values})
+    with ExcelWriter(design_path) as xl_write:
+        design_matrix_df.to_excel(xl_write, index=False, sheet_name="DesignSheet01")
+        DataFrame().to_excel(
```
Comment (jonathan-eq) [inline tests/ert/unit_tests/test_libres_facade.py:271]: Does this create an empty DefaultValues-sheet?

Comment (xjules) [reply tests/ert/unit_tests/test_libres_facade.py:271]: Yes, it does as default is required. But maybe it does not need to be ... :thinking:

Comment (xjules) [reply tests/ert/unit_tests/test_libres_facade.py:271]: We will (for now at least) require both and then we will see if defaultsheet can be removed :+1:

File: src/ert/run_models/ensemble_experiment.py [hunk 9]
```diff
@@ -61,7 +62,42 @@ def run_experiment(
...
+                parameters=parameters_config,
+                observations=self.ert_config.observations,
+                responses=self.ert_config.ensemble_config.response_configuration,
+            )
+            self.ensemble = self._storage.create_ensemble(
+                self.experiment,
+                name=self.ensemble_name,
+                ensemble_size=self.ensemble_size,
+            )
+        elif not restart:
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:100]: Haven't put much thought into restart functionality. Will update!

[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py]: The functionality should be preserved now.

File: src/ert/run_models/ensemble_experiment.py [hunk 10]
```diff
@@ -87,10 +98,14 @@ def run_experiment(
             np.array(self.active_realizations, dtype=bool),
             ensemble=self.ensemble,
         )
+
         sample_prior(
             self.ensemble,
             np.where(self.active_realizations)[0],
             random_seed=self.random_seed,
+            design_matrix_df=(
```
Comment (xjules) [inline src/ert/run_models/ensemble_experiment.py:106]: note to self: This can be simplified

File: src/ert/run_models/ensemble_experiment.py [hunk 11]
```diff
@@ -63,10 +63,26 @@ def run_experiment(
...
+            )
+
+            assert design_matrix.active_realizations is not None
+            self.active_realizations = design_matrix.active_realizations
         if not restart:
             self.experiment = self._storage.create_experiment(
                 name=self.experiment_name,
-                parameters=self.ert_config.ensemble_config.parameter_configuration,
+                parameters=(
+                    [*parameters_config, design_matrix_group]
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:82]: Not anymore, since I didn't want to do storage migration and in this way I can keep it separated.

File: src/ert/gui/simulation/ensemble_experiment_panel.py [hunk 12]
```diff
@@ -85,6 +85,13 @@ def __init__(
 
         design_matrix = analysis_config.design_matrix
         if design_matrix is not None:
+            if design_matrix.design_matrix_df is None:
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/gui/simulation/ensemble_experiment_panel.py:88]: Yes, that's a good one, I will refactor this.

File: src/ert/config/design_matrix.py [hunk 13]
```diff
@@ -74,6 +69,58 @@ def from_config_list(cls, config_list: List[str]) -> "DesignMatrix":
...
+        if self.parameter_configuration is None or not isinstance(
+            self.parameter_configuration[DESIGN_MATRIX_GROUP], GenKwConfig
+        ):
+            return existing_parameters, None
+
+        new_param_config: List[ParameterConfig] = []
+
+        design_parameter_group = self.parameter_configuration[DESIGN_MATRIX_GROUP]
+        design_keys = []
+        if isinstance(design_parameter_group, GenKwConfig):
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/config/design_matrix.py:102]: Mypy requires it.

File: src/ert/gui/simulation/ensemble_experiment_panel.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/gui/simulation/ensemble_experiment_panel.py:88]: Maybe the reading itself should treat the exception and store (if something goes wrong) to a variable and set `desgin_matrix_df=None`

File: src/ert/run_models/ensemble_experiment.py [hunk 14]
```diff
@@ -63,10 +63,26 @@ def run_experiment(
...
+        # to the experiment parameters and set new active realizations
+        parameters_config = self.ert_config.ensemble_config.parameter_configuration
+        design_matrix = self.ert_config.analysis_config.design_matrix
+        design_matrix_group = None
+        if design_matrix is not None:
+            parameters_config, design_matrix_group = (
+                design_matrix.merge_with_existing_parameters(parameters_config)
+            )
+
+            assert design_matrix.active_realizations is not None
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/run_models/ensemble_experiment.py:76]: Yes, it should be

File: src/ert/gui/simulation/ensemble_experiment_panel.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/gui/simulation/ensemble_experiment_panel.py:89]: Maybe moving it to the config validation, but then pd.Dataframe is not serializable. Let's  discuss it with @oyvindeide.

File: src/ert/config/design_matrix.py [hunk 15]
```diff
@@ -76,6 +79,60 @@ def from_config_list(cls, config_list: List[str]) -> "DesignMatrix":
...
+            existing_parameters (List[ParameterConfig]): List of existing parameters
+
+        Raises:
+            ConfigValidationError: If there is a partial overlap between the design matrix group and any existing GEN_KW group
+
+        Returns:
+            tuple[List[ParameterConfig], ParameterConfig]: List of existing parameters and the dedicated design matrix group
+        """
+
+        if self.parameter_configuration is None or not isinstance(
```
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/config/design_matrix.py]: Now it is :+1:

File: src/ert/config/design_matrix.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (xjules) [reply src/ert/config/design_matrix.py:104]: It is still list of `ParameterConfig` so it needs to be there.

File: tests/ert/ui_tests/cli/analysis/test_design_matrix.py [hunk 16]
```diff
@@ -0,0 +1,193 @@
...
+                """
+            )
+        )
+
+    with open("poly_eval.py", "w", encoding="utf-8") as f:
+        f.write(
+            dedent(
+                """\
+                #!/usr/bin/env python
+                import numpy as np
```
[the remark this answers is not in the document]
Comment (xjules) [reply tests/ert/ui_tests/cli/analysis/test_design_matrix.py:55]: No, good catch. it will be removed :+1:

File: tests/ert/unit_tests/config/test_analysis_config.py [hunk 17]
```diff
@@ -18,29 +19,30 @@ def test_analysis_config_from_file_is_same_as_from_dict(monkeypatch, tmp_path):
...
         pass
     monkeypatch.chdir(tmp_path)
-    assert ErtConfig.from_file_contents(
-        dedent(
-            """
-                NUM_REALIZATIONS 10
-                MIN_REALIZATIONS 10
-                ANALYSIS_SET_VAR STD_ENKF ENKF_TRUNCATION 0.8
-                DESIGN_MATRIX my_design_matrix.xlsx DESIGN_SHEET:my_sheet DEFAULT_SHEET:my_default_sheet
+    with suppress(ConfigValidationError):
```
[the remark this answers is not in the document]
Comment (xjules) [reply tests/ert/unit_tests/config/test_analysis_config.py:22]: The test just compares to configs, wherein the assert in the middle does exactly that. Suppress is to neglect validation error from wrongly formatted design matrix.

[the remark this answers is not in the document]
Comment (xjules) [reply tests/ert/unit_tests/config/test_analysis_config.py:22]: right, I sort of thought when the name of test is `test_analysis_config_from_file_is_same_as_from_dict` it only compares the content of the dicts and not go into semantics, but can change it to have a meaningful entries in the matrix and can remove the suppress statement.

File: src/ert/enkf_main.py [hunk 18]
```diff
@@ -162,6 +159,29 @@ def _seed_sequence(seed: Optional[int]) -> int:
     return int_seed
 
 
+def save_design_matrix_to_ensemble(
```
Comment (jonathan-eq) [inline src/ert/enkf_main.py:162]: Should this be a method in the ensemble class? `my_ensemble.save_design_matrix(...)`

Comment (xjules) [reply src/ert/enkf_main.py:165]: There is already `ensemble.save_parameters`, which represents the interface for storing parameters. Not sure if it would bring any value having it there as it is a special "class" of parameters.
